### PR TITLE
WebSocket resume

### PR DESCRIPTION
**Pull request by DedShotTM**

### CODE REVIEW

**Comment by uAliFurkanY** [comment]
image
line 13 throws lint error nodejs undefined
ignore no-undef or fix

**Comment by bsian03** [comment]
Ignore linting errors for now, I am working on a custom eslint configuration and will commit it onto the dev branch when it's complete and working.

File: src/gateway/WebSocket.ts [hunk 1]
```diff
@@ -1,16 +1,34 @@
-/* eslint-disable no-case-declarations */
+/* eslint-disable no-console */
+/* eslint-disable no-use-before-define */
```
Comment (bsian03) [inline src/gateway/WebSocket.ts:2]: Get rid of both of these for now

File: src/gateway/WebSocket.ts [hunk 2]
```diff
@@ -1,16 +1,34 @@
...
+
+  function heartbeat(expected: boolean) {
+    if (expected) {
+      if (!lastHeartbeatAck) {
+        clearInterval(interval);
+        ws.removeEventListener('close');
+        ws.terminate();
+        newWS();
+        resume();
+        lastHeartbeatAck = true;
```
Comment (bsian03) [inline src/gateway/WebSocket.ts:23]: Could probably add a newline somewhere here, preferably after `terminate()`

File: src/gateway/WebSocket.ts [hunk 3]
```diff
@@ -1,16 +1,34 @@
...
+      } else lastHeartbeatAck = false;
+    }
+    if (ws.readyState === 1) {
```
Comment (bsian03) [inline src/gateway/WebSocket.ts:26]: Do we not have a constant for this?

File: src/gateway/WebSocket.ts [hunk 4]
```diff
@@ -21,67 +39,79 @@ export default async function Socket(token: string, wsURL: string, intents: numb
...
         },
       },
     }));
   }
 
-  ws.on('open', () => {
-    identify();
-  });
+  function resume() {
+    if (ws.readyState === 1) {
```
Comment (bsian03) [inline src/gateway/WebSocket.ts:50]: See other constant review

File: src/gateway/WebSocket.ts [hunk 5]
```diff
@@ -21,67 +39,79 @@ export default async function Socket(token: string, wsURL: string, intents: numb
...
 
       case GATEWAY_OPCODES.RECONNECT:
-        break;
-
-      case GATEWAY_OPCODES.REQUEST_GUILD_MEMBERS:
+        clearInterval(interval);
+        ws.removeEventListener('close');
+        ws.terminate();
+        newWS();
+        resume();
```
Comment (bsian03) [inline src/gateway/WebSocket.ts:83]: This could also do with another newline

File: src/gateway/WebSocket.ts [hunk 6]
```diff
@@ -21,67 +39,79 @@ export default async function Socket(token: string, wsURL: string, intents: numb
...
         break;
 
       case GATEWAY_OPCODES.INVALID_SESSION:
+        setTimeout(() => {
+          identify();
+        }, 4000);
         break;
```
Comment (bsian03) [inline src/gateway/WebSocket.ts:90]: You should really check if the session can be resumed. This can be added afterwards if you aren't adding it now

**Review by bsian03 — CHANGES_REQUESTED** [review]

File: src/gateway/WebSocket.ts (hunk 1 above)
Comment (DedShotTM) [reply src/gateway/WebSocket.ts]: done

File: src/gateway/WebSocket.ts (hunk 2 above)
Comment (DedShotTM) [reply src/gateway/WebSocket.ts]: done

File: src/gateway/WebSocket.ts (hunk 4 above)
Comment (DedShotTM) [reply src/gateway/WebSocket.ts]: done

File: src/gateway/WebSocket.ts (hunk 3 above)
Comment (DedShotTM) [reply src/gateway/WebSocket.ts]: yes ws does, done

File: src/gateway/WebSocket.ts (hunk 5 above)
Comment (DedShotTM) [reply src/gateway/WebSocket.ts]: done

File: src/gateway/WebSocket.ts (hunk 6 above)
Comment (DedShotTM) [reply src/gateway/WebSocket.ts:90]: what do you mean exactly

Comment (bsian03) [reply src/gateway/WebSocket.ts:90]: See https://discord.com/developers/docs/topics/gateway#resuming

**Review by bsian03 — APPROVED** [review]
We'll need to convert Gateway over to a class at some point, but not now. We can do that later.

**Review by Khaaz — APPROVED** [review]
This seems to be good.
As mentionned this will need to be switched to a class. Which will also allow us to see more clearly where we are going.
I am not a huge fan of the closures with usage of outter function variables in this case. But this will typically be solved when using a class.

**Pull request merged**